talk/expert/modules.tex

```diff
@@ -0,0 +1,643 @@
+\subsection[modules]{Modules}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{Modules}
+  \begin{block}{Motivation}
+    \begin{itemize}
+      \item Modules allow for sharing declarations and definitions across translation units
+      \item Header files do the same, but modules have benefits:
+      \begin{itemize}
+        \item Better isolation. No cross-talk between multiple pieces of included code, and also not between included code an your code.
+        \item Better control over public interface of your library. Avoid leaking library dependencies into user code.
+        \item Faster compilation
+      \end{itemize}
+    \end{itemize}
+  \end{block}
+  \begin{alertblock}{Textual includes}
+    \begin{cppcode}
+      #include "math.h"
+    \end{cppcode}
+  \end{alertblock}
+  \begin{exampleblock}{Module importation}
+    \begin{cppcode}
+      import math;
+    \end{cppcode}
+  \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{Writing modules}
+  \begin{exampleblock}{ratio.cpp}
+    \begin{cppcode*}{gobble=2}
+      export module math;
+      import <string>;
+      namespace utils { // ns. and module names orthogonal
+        export struct Ratio { float num, denom; };
+        export std::string toString(const Ratio& r) { ... }
+        void normalize(Ratio& r) { ... } // not exported
+      }
+    \end{cppcode*}
+  \end{exampleblock}
+  \begin{exampleblock}{main.cpp}
+    \begin{cppcode*}{gobble=2}
+      import <iostream>;
+      import math;
+      int main() {
+        auto r = utils::Ratio{1, 3};
+        std::cout << utils::toString(r);
+      }
+    \end{cppcode*}
+  \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{Writing modules}
+  \begin{block}{Exporting declarations}
+    \begin{itemize}
+      \item By default, everything inside a module is internal
+      \item Declarations become visible when declared with \mintinline{cpp}{export} in the primary module interface
+      \item You can only export entities at namespace scope
```

Comment:
I had to look up whether the global scope was allowed or not. Maybe somebody else would have had the same question? In that case, we could add:
```suggestion
      \item You can only export entities at namespace scope (including the global scope)
```

Reply:
The global scope is a namespace scope, but I can see how this may be confusing.